### PR TITLE
Initial TCK tests for build compatible extensions

### PR DESCRIPTION
**Pull request by Ladicek**

### CODE REVIEW

**Comment by Ladicek** [comment]
Marked as draft, because I didn't run these tests with any implementation yet. They are, however, a direct port of tests I developed for my Quarkus prototype, so they should mostly be fine.

I'll work on running these tests with an implementation next.

**Comment by Ladicek** [comment]
Amended the commit with a test for synthetic observers of parameterized types.

**Comment by Ladicek** [comment]
Rebased.

As discussed on the CDI call today, we still don't have a way to run this with an implementation, but work is ongoing to run this both with Weld and Quarkus. Once that happens, this PR will be marked as ready for review and assuming no problems are found, merged. More tests need to be added for sure, but those won't block this PR.

File: impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/changeBeanQualifier/ChangeBeanQualifierTest.java
Comment (manovotn) [inline impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/changeBeanQualifier/ChangeBeanQualifierTest.java:16]: This is actually wrong - the archive will have `beans.xml` automatically included (`WebArchiveBuilder` does that).
Instead, we could make use of the `annotated` discovery mode (just the empty beans.xml file) and simply not give those beans any BDA.

File: impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/changeBeanQualifier/MyOtherService.java
Comment (manovotn) [inline impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/changeBeanQualifier/MyOtherService.java:7]: Your deployment assumes no discovery yet this bean is not added via extension.

**Review by manovotn — CHANGES_REQUESTED** [review]
I have glued together several snapshots and I am able to test this with Weld.
Most of the tests are failing, so I am going to look into what's the cause and contribute the changes back here.

Meanwhile I dropped a few comments to things I noticed.

**Comment by manovotn** [comment]
I have added a commit that fixes `ChangeBeanQualifierTest` to demonstrate the issue we described in https://github.com/eclipse-ee4j/cdi/issues/564
@graemerocher ^ just to make clear what we meant :)

**Comment by manovotn** [comment]
Added another commit that fixes all but `SyntheticBeanTest` where I am not quite sure what the issue is yet.

**Comment by manovotn** [comment]
Ok, added another commit that fixes the last test. With these changes I can get Weld's implementation to pass it.

**Comment by manovotn** [comment]
Moved from draft to standard PR.
I will also give it green light from myself but maybe @Ladicek would like to review it before merging since I made quite a few changes :)

**Review by manovotn — APPROVED** [review]

File: impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticBean/SyntheticBeanExtension.java
Comment (Ladicek) [inline impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticBean/SyntheticBeanExtension.java:33]: ??? :-)

File: impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/customStereotype/MyService.java
Comment (Ladicek) [inline impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/customStereotype/MyService.java:3]: Why is this required? The class already has a stereotype?

Comment (manovotn) [reply impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/customStereotype/MyService.java:3]: Originally, `MyService` was registered via `@Discovery` manually, I simply made sure it will be replaced by natural discovery process.

**Comment by Ladicek** [comment]
Hi @manovotn, I've pushed a few more commits with some proposed changes, so that:

- the tests pass on ArC;
- multiple combinations of `beans.xml` present/missing and classes added/not added via `@Discovery` are tested;
- this still follows the rules for type discovery as we discussed.

If you can let me know how I can test with Weld, I'd be happy to do that.

**Comment by manovotn** [comment]
@Ladicek I am seeing one failing test with Weld. `SyntheticObserverOfParameterizedTypeExtension` is adding `MyService` but that class also contains a bean defining annotation and as such is discovered. 

The way to execute it with Weld is:
* Check out master branch
* Perform a `mvn clean install -DskipTests`
* In `core/pom.xml`, change the value of `<cdi.tck-4-0.version>` property to `4.0.0-SNAPSHOT`
  * This OFC assumes you have CDI TCK built off this PR
* Run `mvn clean verify -f jboss-tck-runner/pom.xml`

File: impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticObserverOfParameterizedType/SyntheticObserverOfParameterizedTypeExtension.java
Comment (manovotn) [inline impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticObserverOfParameterizedType/SyntheticObserverOfParameterizedTypeExtension.java:15]: @Ladicek This is redundant and causes failures for Weld.
`MyService` has `@ApplicationScoped` and as such will be discovered.
Will you fix that or should I?

File: web/src/main/java/org/jboss/cdi/tck/shrinkwrap/ee/WebArchiveBuilder.java
Comment (manovotn) [inline web/src/main/java/org/jboss/cdi/tck/shrinkwrap/ee/WebArchiveBuilder.java:97]: Do we actually need to have two `WebArchiveBuilder` classes? There is one in `impl` and other under `web`; seems like a remnant of the TCK split.

Comment (starksm64) [reply web/src/main/java/org/jboss/cdi/tck/shrinkwrap/ee/WebArchiveBuilder.java:97]: The one in in the web module has an ejb module that is included by default in test deployments and has a dependency on the EnterpriseArchiveBuilder.

File: impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticObserverOfParameterizedType/SyntheticObserverOfParameterizedTypeExtension.java
Comment (Ladicek) [reply impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticObserverOfParameterizedType/SyntheticObserverOfParameterizedTypeExtension.java:15]: I would call that a bug _somewhere_. I intentionally added 1 or 2 cases where `@Discovery` is used to add a bean which is discovered through normal type discovery. I'm not aware of anything in the spec that would say this is illegal, so if that's what should happen in your opinion, we need to specify that somewhere.

File: web/src/main/java/org/jboss/cdi/tck/shrinkwrap/ee/WebArchiveBuilder.java
Comment (Ladicek) [reply web/src/main/java/org/jboss/cdi/tck/shrinkwrap/ee/WebArchiveBuilder.java:97]: Perhaps so, I didn't really check.

File: impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticObserverOfParameterizedType/SyntheticObserverOfParameterizedTypeExtension.java
Comment (manovotn) [reply impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticObserverOfParameterizedType/SyntheticObserverOfParameterizedTypeExtension.java:15]: @Ladicek Yes you can do that. The discovered and added types (based off the same java class) are two different ones but with equal bean types and qualifiers (unless changed later on in the bootstrap OFC).
You should be getting an ambiguous resolution exception.

Taking portable extensions as an example - if you take a look at BBD javadoc, it has a specific parameter that serves this purpose - https://github.com/eclipse-ee4j/cdi/blob/master/api/src/main/java/jakarta/enterprise/inject/spi/BeforeBeanDiscovery.java#L162
In fact, you could in theory add one AT several times with just different IDs.

Comment (Ladicek) [reply impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticObserverOfParameterizedType/SyntheticObserverOfParameterizedTypeExtension.java:15]: My expectation actually was that the container would ignore the duplicate :-) We need to specify this somehow.

**Comment by Ladicek** [comment]
Added one commit to avoid the type discovery duplicate.

File: impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticObserverOfParameterizedType/SyntheticObserverOfParameterizedTypeExtension.java
Comment (Ladicek) [reply impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticObserverOfParameterizedType/SyntheticObserverOfParameterizedTypeExtension.java]: This was specified in https://github.com/eclipse-ee4j/cdi/pull/571 (as non-portable) and the test was corrected, so resolving this open question.

**Comment by Ladicek** [comment]
I think this is ready to be merged now, because it passes on ArC and Weld. I'll just restructure the commits a little and squash a bunch of them.

**Comment by Ladicek** [comment]
Done.

**Review by graemerocher — APPROVED** [review]

**Review by manovotn — APPROVED** [review]

**Pull request merged**
